pkg/envoy/xds/watcher.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/envoy/api"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/sirupsen/logrus"
+)
+
+// ResourceWatcher watches and retrieves new versions of resources from a
+// resource set.
+// ResourceWatcher implements ResourceVersionListener to get notified when new
+// resource versions are available in the set.
+type ResourceWatcher struct {
+	// typeURL is the URL that uniquely identifies the resource type.
+	typeURL string
+
+	// resourceSet is the set of resources to watch.
+	resourceSet ResourceSet
+
+	// version is the current version of the resources. Updated in calls to
+	// NotifyNewVersion.
+	version uint64
+
+	// versionLocker is used to lock all accesses to version.
+	versionLocker lock.Mutex
+
+	// versionCond is a condition that is broadcast whenever the source's
+	// current version is increased.
+	// versionCond is associated with versionLocker.
+	versionCond *sync.Cond
+
+	// resourceAccessTimeout is the timeout to use for any access to the
+	// resource set.
+	resourceAccessTimeout time.Duration
+}
+
+// NewResourceWatcher creates a new ResourceWatcher backed by the given
+// resource set.
+func NewResourceWatcher(typeURL string, resourceSet ResourceSet, resourceAccessTimeout time.Duration) *ResourceWatcher {
+	w := &ResourceWatcher{
+		typeURL:               typeURL,
+		resourceSet:           resourceSet,
+		resourceAccessTimeout: resourceAccessTimeout,
+	}
+	w.versionCond = sync.NewCond(&w.versionLocker)
+	return w
+}
+
+func (w *ResourceWatcher) HandleNewResourceVersion(typeURL string, version uint64) {
+	w.versionLocker.Lock()
+	defer w.versionLocker.Unlock()
+
+	if typeURL != w.typeURL {
+		return
+	}
+
+	if version < w.version {
+		panic(fmt.Sprintf("decreasing version number found for resources of type %s: %d < %d",
```

Comment:
Version numbers are generated by the implementations of the `ResourceSet` interface within the Cilium agent itself. It's part of the contract of that interface that version numbers are monotonically increasing. If an implementation doesn't respect that contract, it's a serious bug in the Cilium agent. Nothing to do with protocol clients.

Reply:
Fair enough. We already have panics in the code, it seems. My general aversion to panics is that they undermine the programs ability to report errors. You'll get something in the logs, certainly, but this then requires a log aggregation system and a supervisor to restart. Both are common but log aggregators create other problems (more than they solve, in my experience). Anyway, this panic is quite unlikely so carry on!

Reply:
@raybejjani would calling `log.Panicf` be better?

Reply:
Oh, good idea. I guess that would be more consistent with the other logs, since the panic message will have the same field names. If it's the same to you then I'd vote for using `log.panicf`.

Reply:
Done!